harness/determined/common/storage/boto3_credential_manager.py

```diff
@@ -0,0 +1,148 @@
+    Creates a refreshable credential provider class given an existing credential provider in
+    the boto3 credential chain.
+
+    """
+
+    METHOD = "managed-refresh-cred"
+
+    def __init__(self, credential_provider: SharedCredentialProvider, check_every: int = 2) -> None:
+        super().__init__()
+        self.check_every = timedelta(minutes=check_every)
+        self.credential_provider = credential_provider
+        self.last_loaded = self._credentials_modified_time()
+
+    def load(self) -> Optional[Credentials]:
+        credentials = self._load_credentials()
+        return credentials and RefreshedSharedCredentials(
+            refresh_using=self._load_credentials,
+            refresh_needed=self._reload_needed,
+            check_every=self.check_every,
+            access_key=credentials.access_key,
+            secret_key=credentials.secret_key,
+            token=credentials.token,
+        )
+
+    def _reload_needed(self) -> bool:
+        return self._credentials_modified_time() > self.last_loaded
+
+    def _load_credentials(self) -> Credentials:
+        self.last_loaded = self._credentials_modified_time()
+        return self.credential_provider.load()
+
+    def _credentials_file(self) -> Any:
+        path = self.credential_provider._creds_filename
+        path = os.path.expandvars(path)
+        path = os.path.expanduser(path)
+        return path
+
+    def _credentials_modified_time(self) -> float:
+        credentials_file = self._credentials_file()
+        return os.stat(credentials_file).st_mtime
+
+
+class RefreshedSharedCredentials(Credentials):  # type: ignore
+    def __init__(
+        self,
+        access_key: str,
+        secret_key: str,
+        token: str,
+        check_every: timedelta,
+        refresh_using: Callable,
+        refresh_needed: Callable,
+    ):
+        self._refresh_using = refresh_using
+        self._refresh_needed = refresh_needed
+        self._access_key = ensure_unicode(access_key)
+        self._secret_key = ensure_unicode(secret_key)
+        self._token = token
+        self._check_every = check_every
+        self._lock = threading.Lock()
+        self._check_time = self._next_check_time()
+        self._frozen_credentials = ReadOnlyCredentials(access_key, secret_key, token)
+
+    def _refresh(self) -> None:
+        now = _local_now()
+        if now < self._check_time:
+            return
+        with self._lock:
+            if now < self._check_time:
+                return
+            self._check_time = self._next_check_time()
+            if self._refresh_needed():
+                logging.info("credential file changes detected, refreshing credentials")
+                credentials = self._refresh_using()
+                self.access_key = credentials.access_key
+                self.secret_key = credentials.secret_key
+                self.token = credentials.token
+
+    def _next_check_time(self) -> Any:
+        return _local_now() + self._check_every
```

Comment:
This is a bit of a style preference thing, not a blocker or anything, but I think this helper method is too small to be helpful.

If I read:
```python
self._check_time = self._next_check_time()
```
I have to go look at `self._next_check_time()` to know what it does.

If I read:
```python
self._check_time = time.time() + self._backoff_time
```
I know immediately what it does.  And `time.time() + self._backoff_time` is so simple and easy I don't feel like it needs to have a helper function for the two times you would use it.

Helper functions aren't zero-cost; when used excessively they make simple code which could be understood linearly no longer possible to understand linearly (you have to jump around reading helper functions instead).

Again, this is style thing more than a hard blocker.  This is just my take.